includes/class-wcpdf-admin.php

```diff
@@ -1154,6 +1040,9 @@ public function ajax_crud_document() {
 		);
 
 		try {
+			// Allow getting documents without invoice.
+			add_filter( 'wpo_wcpdf_allow_document_without_invoice', '__return_true' );
```

Comment:
Maybe this could be done differently? For example a filter with an array of Invoice depending document types? Also you don't offer any variable to check which document is affected by this filter, so maybe this should be after the line `1046`.

Reply:
> Maybe this could be done differently? For example a filter with an array of Invoice depending document types?

I'm sorry, I couldn't fully understand your approach. Could you please explain more?
I added this filter here so third parties, such as our Pro extension, can use it to affect the result of the `is_allow()` method. Do you mean to change [this filter](https://github.com/wpovernight/woocommerce-pdf-ips-pro/pull/392/files#diff-1aaf16af4dbfbd03af9365d3d5500e1e5e75a84b1e49aa025a71eb5b38875578R1366) from boolean to an array containing depended documents, and then we make that array empty in the free plugin?

Reply:
> Do you mean to change [this filter](https://github.com/wpovernight/woocommerce-pdf-ips-pro/pull/392/files#diff-1aaf16af4dbfbd03af9365d3d5500e1e5e75a84b1e49aa025a71eb5b38875578R1366) from boolean to an array containing depended documents, and then we make that array empty in the free plugin?

Yes.

Reply:
I updated that hook here: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/pull/757/commits/55b47a8df4af86bbad9bd30a21d01e3e1a458768, and also in the Pro branch.

> Also you don't offer any variable to check which document is affected by this filter, so maybe this should be after the line 1046.

I couldn't understand this. Could you please explain?

Reply:
The implementation seems a bit strange to me, because the filter has to be placed in Pro when should be the contrary.

Reply:
I think I saw this reverse use of filters a few times, maybe even in WooCommerce. This is one way we can inform third parties that we need to get the document anyway.

Another implementation that I could think of is to declare a new property for invoice requirement, for example, a boolean `$invoice_required` property in the `Order_Document` class, and a method like this inside this class:

```php
public function is_invoice_required(): bool {
	return apply_filters( 'wpo_wcpdf_document_is_invoice_required', $this->invoice_required, $this );
}
```

So, this method could be used inside the `is_pro_document_allowed()` method or by third parties, and we can change it using the provided filter to display the document in the meta box.

Or we can also generalize this by changing the `$invoice_required` property to an array property `$prerequisite_documents`, and change that method to something like this:
```php
public function get_prerequisites(): array {
	return apply_filters( 'wpo_wcpdf_document_prerequisites', $this->$prerequisite_documents, $this );
}
```

What do you think?

Reply:
WPML is one of those examples, but if we go that way we need a proper documentation about that and also custom documents.

Reply:
I agree that this reverse use of filters might be confusing and difficult for third parties to understand. So, let's go with the last method.

Among the options for name, I think both `get_prerequisite_documents()` and `get_base_documents()` are good.
Since this requirement means the document can exist but not be displayed, I thought maybe the 'prerequisite' could be a good word.
Which one do you prefer?

Reply:
Your suggestion `get_prerequisite_documents` is good.

Reply:
Updated with the new implementation: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/pull/757/commits/a6c59279b703c8ab239f0f297dfa727037e53ba8.